tests/system/test_pandas.py

```diff
@@ -0,0 +1,794 @@
+        for col in column_names:
+            # all the schema fields are nullable, so None is acceptable
+            if not row[col] is None:
+                assert isinstance(row[col], exp_datatypes[col])
+
+
+def test_insert_rows_from_dataframe(bigquery_client, dataset_id):
+    SF = bigquery.SchemaField
+    schema = [
+        SF("float_col", "FLOAT", mode="REQUIRED"),
+        SF("int_col", "INTEGER", mode="REQUIRED"),
+        SF("bool_col", "BOOLEAN", mode="REQUIRED"),
+        SF("string_col", "STRING", mode="NULLABLE"),
+    ]
+
+    dataframe = pandas.DataFrame(
+        [
+            {
+                "float_col": 1.11,
+                "bool_col": True,
+                "string_col": "my string",
+                "int_col": 10,
+            },
+            {
+                "float_col": 2.22,
+                "bool_col": False,
+                "string_col": "another string",
+                "int_col": 20,
+            },
+            {
+                "float_col": 3.33,
+                "bool_col": False,
+                "string_col": "another string",
+                "int_col": 30,
+            },
+            {
+                "float_col": 4.44,
+                "bool_col": True,
+                "string_col": "another string",
+                "int_col": 40,
+            },
+            {
+                "float_col": 5.55,
+                "bool_col": False,
+                "string_col": "another string",
+                "int_col": 50,
+            },
+            {
+                "float_col": 6.66,
+                "bool_col": True,
+                # Include a NaN value, because pandas often uses NaN as a
+                # NULL value indicator.
+                "string_col": float("NaN"),
+                "int_col": 60,
+            },
+        ]
+    )
+
+    table_id = f"{bigquery_client.project}.{dataset_id}.test_insert_rows_from_dataframe"
+    table_arg = bigquery.Table(table_id, schema=schema)
+    table = helpers.retry_403(bigquery_client.create_table)(table_arg)
+
+    chunk_errors = bigquery_client.insert_rows_from_dataframe(
+        table, dataframe, chunk_size=3
+    )
+    for errors in chunk_errors:
+        assert not errors
+
+    # Use query to fetch rows instead of listing directly from the table so
+    # that we get values from the streaming buffer.
+    rows = list(
+        bigquery_client.query(
+            "SELECT * FROM `{}.{}.{}`".format(
+                table.project, table.dataset_id, table.table_id
+            )
+        )
+    )
+
+    sorted_rows = sorted(rows, key=operator.attrgetter("int_col"))
+    row_tuples = [r.values() for r in sorted_rows]
+    expected = [
+        tuple(None if col != col else col for col in data_row)
+        for data_row in dataframe.itertuples(index=False)
+    ]
+
+    assert len(row_tuples) == len(expected)
+
+    for row, expected_row in zip(row_tuples, expected):
+        assert len(row) == len(expected_row)  # column order does not matter
```

Comment:
This assertion is weaker than the original - [assertCountEqual()](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual) not only checks that the lengths of two sequences match, but also that the counts of their individual elements match.

If the elements are hashable, the easiest way to make an equivalent assertion is probably using the `collections.Counter` class:
```py
assert Counter(row) == Counter(expected_row)
```

Reply:
Nice catch!